sw/device/tests/entropy_src_fw_override_test.c

```diff
@@ -0,0 +1,407 @@
+   */
+  kEntropySrcFifoThreshold = 32,
+  /**
+   * The number of contiguous samples we want to capture.
+   */
+  kContiguousSamplesCount = 1024,
+  /**
+   * Number of bits per sample.
+   */
+  kBitsPerSample = 4,
+  /**
+   * Timeout for operations which wait for random data.
+   */
+  kTimeoutUsec = 50000,
+  /**
+   * Size of buffer in words to hold all the samples, assuming
+   * 4-bit samples at the most.
+   */
+  kFifoBufferSizeWords =
+      kContiguousSamplesCount * kBitsPerSample / sizeof(uint32_t),
+};
+
+static dif_aes_t aes;
+static dif_csrng_t csrng;
+static dif_edn_t edn0;
+static dif_edn_t edn1;
+static dif_entropy_src_t entropy_src;
+static dif_hmac_t hmac;
+static dif_kmac_t kmac;
+static dif_rv_core_ibex_t rv_core_ibex;
+static dif_rv_plic_t rv_plic;
+
+// Buffers to collect samples into.
+//
+// We use a double-buffering approach, where one buffer is for the "input"
+// samples that we read from the observe FIFO, and the other is for "output"
+// samples that we need to write to the conditioner.
+//
+// When the input buffer is filled and the output buffer is empty, we swap them
+// and the bytes we collected are written out to the conditioner. The input
+// buffer is now empty, and we can start writing new samples to it.
+static uint32_t sample_buf_0[kFifoBufferSizeWords];
+static uint32_t sample_buf_1[kFifoBufferSizeWords];
+
+// These pointers track which buffer is the input or the output and are cursors
+// into where we have written to / read from them so far.
+static uint32_t *input_buf = sample_buf_0;
+static uint32_t *output_buf = sample_buf_1 + kFifoBufferSizeWords;
+
+// These counters track how many words are remaining to be written to the input
+// buffer or read out of the output buffer.
+static size_t words_to_input = kFifoBufferSizeWords;
+static size_t words_to_output = 0;
+
+static dif_edn_auto_params_t edn_params0;
+static dif_edn_auto_params_t edn_params1;
+
+/**
+ * Determine whether the observe FIFO has overflowed.
+ *
+ * TODO(#21279) Normally, one would rely on the FW_OV_RD_FIFO_OVERFLOW
+ * register but due to an RTL bug, the overflow bit is pulsed
+ * instead of latched so we cannot rely on it. Instead, rely
+ * on OBSERVE_FIFO_DEPTH and assume that if the FIFO is full
+ * then it has overflowed.
+ */
+bool entropy_src_fifo_has_overflowed(void) {
+  uint32_t fifo_depth;
+  CHECK_DIF_OK(dif_entropy_src_get_fifo_depth(&entropy_src, &fifo_depth));
+  return fifo_depth == ENTROPY_SRC_PARAM_OBSERVE_FIFO_DEPTH;
+}
+
+/**
+ * Drain observe FIFO.
+ */
+static void drain_observe_fifo(void) {
+  // This value is arbitrary, it could be 1 but since there is some
+  // overhead in dif_entropy_src_observe_fifo_nonblocking_read, it's better
+  // to read several words every time to drain the FIFO quickly.
+  const size_t kDrainCount = 32;
+  size_t len;
+  // Read from the FIFO until we get a short read which means that the FIFO was
+  // emptied.
+  LOG_INFO("drain observe FIFO overflow...");
```

Comment:
probably should not log in this function because it can be called in ISR context

Reply:
If you really do want to report that an overflow occured then maybe have a global variable counting the stats (e.g. number of overflows) and display that at the end of the test?

Reply:
I've moved the log outside of the function, but haven't added a global counter even though that could be useful for debugging this test.